### ISSUE TITLE
Defining A Reliably Unequal Type?

### ISSUE THREAD

**Issue by vexx32** [PowerShell/PowerShell#9435]
# Support Question

I have a somewhat unusual problem. In [PSKoans](https://github.com/vexx32/pskoans), I need to define a type such that no matter what you try to compare it to, the result is always `$false`. This is needed to be a blank placeholder that _always_ fails Pester tests with comparisons against various items.

Initially, my attempt looked like this:

```powershell
class Blank {
    [string] ToString() {
        return $null
    }

    static [bool] op_Equality([Blank] $self, [object] $other) {
        return $false
    }

    static [bool] op_Inequality([Blank] $self, [object] $other) {
        return $true
    }

    static [bool] op_Explicit([Blank] $Instance) {
        return $false
    }

    static [bool] op_Implicit([Blank] $Instance) {
        return $false
    }
}
```

This works for most things, but has a pretty glaring issue. The comparison `$true -eq [Blank]::new()` returns `$true`. As a result, a Pester assertion such as `[Blank]::new() | Should -BeTrue` _passes_, when it shouldn't. 

A quick second attempt using a C# struct inlined with Add-Type looks like this:

```powershell
Add-Type -TypeDefinition @"
public struct Blank
{
    public static implicit operator bool(Blank item) => false;

    public static bool operator ==(Blank item, object other) => false;

    public static bool operator !=(Blank item, object other) => false;

    public override bool Equals(object o) => false;

    public override int GetHashCode()
    {
        return 0;
    }
}
"@
```

This exhibits the same issue.

Is there a way that I can define such a type in PowerShell or in C# that will function as desired within PowerShell?

**Comment by mklement0** [comment]
With `$true` as the _LHS_ PowerShell seemingly never consults the methods / interfaces of the RHS and considers _any_ (non-primitive, non-collection) type instance to be implicitly `$true`, using its built-in to-Boolean conversion logic.

```powershell
$true -eq [Blank]::new()
```

is the same as:

```powershell
$true -eq [bool] [Blank]::new()
```

or simply:

```powershell
[bool] [Blank]::new()  # always $true for any non-primitive, non-collection type instance
```

https://github.com/PowerShell/PowerShell/blob/0bd93777fceab06b0b3d949a5d47cd39cf25d65d/src/System.Management.Automation/engine/LanguagePrimitives.cs#L1638-L1644 seems to confirm that.

In short: with a Boolean as the _LHS_, it is only ever PowerShell's built-in to-Boolean conversions that apply.

With the Boolean on the _RHS_, your `[Blank]` type doesn't need to do any extra work to always compare as `$false`:

```powershell
class Blank { }
[Blank]::new() -eq $true # $false
[Blank]::new() -eq $false # $false
```

Aside from that, it seems that even with the operands reversed your static `op_Equality()` method wouldn't be called by `-eq`; you'd need an instance method `[bool] Equals([object] $o) { return $false }`, which would then be called by both `-eq` and `-ne`.

**Comment by vexx32** [comment]
I've been talking with @JamesWTruher in the PS Slack space on some possibilities here, but it looks like as long as the LHS is a boolean, it's impossible to override the equality checks in a way that PS will recognise.

I'm working around this by adjusting the assertion, but I appreciate nonetheless the point of the `Equals()` instance method as it does allow me to have `$a -eq $a` return `$false` which I'm sure I'll run into needing sooner or later!

Closing this as a resounding "no" for wherever booleans are concerned, but in all other cases it looks like I can more or less make it happen. 😄

**Comment by MatejKafka** [comment]
I'm curious, is there a way to do it if LHS is not a boolean? I've been experimenting with a string on the LHS and a custom type on the RHS, and I can't seem to be able to make PowerShell use my custom comparison between the two. I tried operator overloads, `IComparable` and `IComparable<string>`, and PowerShell seems to ignore all three and instead converts RHS to a string.
